Added assertEqualsFloat to qx.core.Assert and qx.core.MAssert

mbgonicus opened the pull request:
One thing I really dislike about computers is that their memory and calculation precision is naturally finite. Of course I dropped that brick when I recently wanted to assert that two floats are equal. Couldn't find a function in the qooxdoo framework that allows an assertion of the equality of two floats respecting a precision, thus I wrote it on my own. Hope you like it.

mbgonicus committed:
Added assertEqualsFloat to qx.core.Assert and qx.core.MAssert

oetiker commented on framework/source/class/qx/core/Assert.js:
```diff
@@ -198,6 +198,33 @@ qx.Bootstrap.define("qx.core.Assert",
+      this.assertNumber(expected);
+      this.assertNumber(found);
+      if (!qx.lang.Type.isNumber(precision)) {
+        precision = 0.01;
```
this will not work well for small or very large numbers ... there are established algorithms for this kind of problem ...I can look them up for you if you wish

oetiker replied:
http://mathjs.org/docs/datatypes/numbers.html

Equality#

Because of rounding errors in calculations, it is unsafe to compare JavaScript Numbers. For example executing 0.1 + 0.2 == 0.3 in JavaScript will return false, as the addition 0.1 + 0.2 introduces a round-off error and does not return exactly 0.3.

To solve this problem, the relational functions of math.js check whether the relative difference between the compared values is smaller than the configured option epsilon. In pseudo code (without exceptions for 0, Infinity and NaN):

diff = abs(x - y)
nearlyEqual = (diff <= max(abs(x), abs(y)) * EPSILON) OR (diff < DBL_EPSILON)
where:

EPSILON is the relative difference between x and y. Epsilon is configurable and is 1e-14 by default. See Configuration.
DBL_EPSILON is the minimum positive floating point number such that 1.0 + DBL_EPSILON != 1.0. This is a constant with a value of approximately 2.2204460492503130808472633361816e-16;
Note that the relational functions cannot be used to compare small values (< 2.22e-16). These values are all considered equal to zero.

oetiker replied:
http://mathjs.org/docs/datatypes/numbers.html

Equality#

Because of rounding errors in calculations, it is unsafe to compare JavaScript Numbers. For example executing 0.1 + 0.2 == 0.3 in JavaScript will return false, as the addition 0.1 + 0.2 introduces a round-off error and does not return exactly 0.3.

To solve this problem, the relational functions of math.js check whether the relative difference between the compared values is smaller than the configured option epsilon. In pseudo code (without exceptions for 0, Infinity and NaN):

diff = abs(x - y)
nearlyEqual = (diff <= max(abs(x), abs(y)) * EPSILON) OR (diff < DBL_EPSILON)
where:

EPSILON is the relative difference between x and y. Epsilon is configurable and is 1e-14 by default. See Configuration.
DBL_EPSILON is the minimum positive floating point number such that 1.0 + DBL_EPSILON != 1.0. This is a constant with a value of approximately 2.2204460492503130808472633361816e-16;
Note that the relational functions cannot be used to compare small values (< 2.22e-16). These values are all considered equal to zero.

oetiker replied:
there is even Number.EPSILON defined in Many js implementations

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/EPSILON

mbgonicus committed:
Fixed algorithm for assertEqualsFloat

mbgonicus committed:
Added qx.lang.Number.equals

mbgonicus committed:
Using qx.lang.Number.equals for assertEqualsFloat

mbgonicus committed:
Added assertNotEqualsFloat

mbgonicus commented:
@oetiker Of course you are right. In a framework this should be done properly. 

I committed a couple of changes:
* introduced qx.lang.Number.equals using the above stated algorithm
* using qx.lang.Number.equals inside qx.core.Assert.assertEqualsFloat
* added qx.core.Assert.assertNotEqualsFloat

oetiker commented on framework/source/class/qx/lang/Number.js:
```diff
@@ -74,6 +74,27 @@ qx.Class.define("qx.lang.Number",
+
+      // 1e-14 is the relative difference
+      return x === y ||
+             Math.abs(x - y) < Number.EPSILON ||
```
not all browsers seem to support EPSILON, using 2^-52 as a default is suggested
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/EPSILON

mbgonicus committed:
Using 2e-52 instead of Number.EPSILON due to lacking browser support

oetiker commented on framework/source/class/qx/lang/Number.js:
```diff
@@ -91,9 +91,12 @@ qx.Class.define("qx.lang.Number",
+      // 1e-14 is the relative difference.
       return x === y ||
-             Math.abs(x - y) < Number.EPSILON ||
+             Math.abs(x - y) < 2e-52 ||
```
aah :) sorry ... what I meant is to use 2e-52 IF EPSILON was not defined

johnspackman replied:
Could we implement Number.EPSILON as a shim then, in a new `qx.lang.normalize.Number`?

mbgonicus replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/EPSILON suggests a simple polyfill:

```javascript
if (Number.EPSILON === undefined){
    Number.EPSILON  =  Math.pow(2, -52);
```

But if Number.EPSILON is 2e-52, then why can't we just use the literal constant?

oetiker replied:
since most browsers support it ... so there is no point in ignoring it ...

mbgonicus committed:
Using Number.EPSILON in qx.lang.Number#equals if available

oetiker commented on framework/source/class/qx/lang/Number.js:
```diff
@@ -74,6 +74,32 @@ qx.Class.define("qx.lang.Number",
+     * @param y {Number}
+     * @return {Boolean}
+     */
+    equals : function(x, y)
```
Sorry to be a pain ... but I think such a call should be called `almostEqual`

at least via 
https://www.google.ch/search?num=100&newwindow=1&dcr=0&q=almost+equal
you find  many other attempts to solve that problem

mbgonicus replied:
I disagree. This is actually to check if two numbers are equal. It just tries to bypass the inefficiencies of a computer to precisely represent float numbers.

It becomes obvious for the example that 0.1 + 0.2 = 0.3. Both sides of that equation are equal, not "almost" equal.

btw: I'd really prefer if you put your criticism here at once and not in dribs and drabs.

oetiker commented on framework/source/class/qx/test/core/Assert.js:
```diff
@@ -153,6 +153,49 @@ qx.Class.define("qx.test.core.Assert",
       }, qx.core.AssertionError);
+    },
+
+    testAssertEqualsFloat : function()
```
same problem

oetiker commented on framework/source/class/qx/test/lang/Number.js:
```diff
@@ -0,0 +1,39 @@
+
+  members :
+  {
+    testEquals : function() {
```
same problem

oetiker commented on framework/source/class/qx/core/MAssert.js:
```diff
@@ -99,6 +99,33 @@ qx.Mixin.define("qx.core.MAssert",
+     * @param found {Float} Found value
+     * @param msg {String} Message to be shown if the assertion fails.
+     */
+    assertEqualsFloat : function(expected, found, msg) {
```
see further down

oetiker commented on framework/source/class/qx/core/Assert.js:
```diff
@@ -198,6 +198,44 @@ qx.Bootstrap.define("qx.core.Assert",
+     * @param found {Float} Found value
+     * @param msg {String} Message to be shown if the assertion fails.
+     */
+    assertEqualsFloat : function(expected, found, msg)
```
see below

mbgonicus committed:
Added polyfill for Number.EPSILON; use that in qx.lang.Number.equals

mbgonicus commented:
I added a polyfill for Number.EPSILON. Thanks to @cajus !

level420 commented:
More than 2 business days, 3 approvals, merging
thank you @mbgonicus
